### PR TITLE
RFC: Browser Support Minbar for v9

### LINKED ISSUE #22273: Determine browser minbar for v9 from current master
We need to determine what the minimum browser version minbars are for Fluent UI v9. This will involve going through our code base and the codebases of all v9 dependencies and analyze:

* ECMAscript syntax and APIs
* Browser platform APIs (i.e. ResizeObserver, MutationObserver, CustomEvent)
* CSS APIs

We can use https://caniuse.com/ to check what are the oldest versions for each of our supported browsers that we can use.

If possible, try to do this with automation so that we can regularly check what our minbar is and if we ever exceed it by accident. Typescript lib.dom types are tied with ecmascript versions which is not a completely accurate mapping to browser supported features.

### PR DESCRIPTION
**Pull request by TristanWatanabe**
This RFC proposes a browser minbar support matrix for v9 to conform to. To get breakdown that led to this matrix, please read this [document](https://microsoft.sharepoint-df.com/teams/FluentUIInternal/_layouts/15/Doc.aspx?sourcedoc=%7B66e5f1ff-23d0-49ab-9a7d-acf3b47344f3%7D&action=edit&wd=target(Build%20system.one%7C13440648-7e78-4299-bc71-661d59756ff2%2FV9%20Browser%20Minbar%20Conclusion%7Cfe9116b5-693e-4559-aaf2-27ca7725abd2%2F)&share=IgH_8eVm0COrSZp9rPO0c0TzAZ1rnQYDmLb_8iCSF8BicSQ).

[Preview👀](https://github.com/microsoft/fluentui/pull/22889/files?short_path=6386842#diff-638684245ec063dc55bc05729c53b47175161565a02cdf30fb334a8a59fbf863)

Current global browser coverage of proposed browser matrix: **88%**
- https://browserslist.dev/?q=Y2hyb21lID49IDg0LCBmaXJlZm94ID49IDc1LCBlZGdlID49IDg0LCBzYWZhcmkgPj0gMTQuMSwgb3BlcmEgPj0gNzMsIGlvc19zYWYgPj0gMTQuNSwgQ2hyb21lQW5kcm9pZCA%2BPSA4NCwgc2Ftc3VuZyA%2BPTE2

## Related Issue(s)

Fixes #22273

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/447a9aa0c3527d42bd071e63777b7afa1a14b273'>447a9aa0c3527d42bd071e63777b7afa1a14b273</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by ling1726** [comment]
Should we put in place some kind of audit process (i.e. every year/6 months) to review and update this browser support matrix ?

**Comment by spmonahan** [comment]
> Should we put in place some kind of audit process (i.e. every year/6 months) to review and update this browser support matrix ?

Maybe sync it with the [Firefox ESR release calendar](https://support.mozilla.org/en-US/kb/firefox-esr-release-cycle) or Safari releases?

**Comment by TristanWatanabe** [comment]
> > Should we put in place some kind of audit process (i.e. every year/6 months) to review and update this browser support matrix ?
> 
> Maybe sync it with the [Firefox ESR release calendar](https://support.mozilla.org/en-US/kb/firefox-esr-release-cycle) or Safari releases?

Went with basing it on Safari major releases since that browser is the new IE

**Review by ling1726 — APPROVED** [review]

File: rfcs/shared/build-system/06-browser-support-for-v9.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+### Pros
+
+- A source of truth for Fluent team and consumers to refer to for browser support.
+- Able to ship modern ES2019 code with the option of going to ES2020 without downleveling.
+- Less maintanance by putting onus of older/unsupported browsers to consumers.
+
+### Cons
+
+- Incompatible CSS features (`flex gap`) are already present in the library and will need to be removed and refactored.
```
Comment (layershifter) [inline rfcs/shared/build-system/06-browser-support-for-v9.md:55]: I would also add [`:focus-visible`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible) to the list. But we have already used `keyborg` and [its custom selector](https://github.com/microsoft/fluentui/blob/8afe4d15242d6220f202ca1d0d1522f4f9d151bb/packages/react-components/react-tabster/src/focus/createCustomFocusIndicatorStyle.ts#L22).

Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:55]: Thanks for pointing this out! Will add that to RFC and from a quick search, we also use this in `react-input`, `react-select`, `react-spinner`, `react-tabster` and `react-textarea`.

File: rfcs/shared/build-system/06-browser-support-for-v9.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+As it stands, Fluent v9 does not currently adhere to a specific browser support matrix. This makes it difficult for partner teams and general consumers of the library to know how using the library will impact their own product. There's also a lack of guidance for the internal team in terms of the syntax, browser APIs and CSS features that they can/can't use.
+
+## Detailed Design or Proposal
+
+### Browser Support Matrix
+
+The proposed browser support matrix will be loosely based off the [Resize Observer](https://caniuse.com/mdn-api_resizeobserver) support matrix . This decision was driven by the current usage of the `ResizeObserver` in Fluent v9 as well as the current business goals of our partner teams. Along with that, being able to maintain our current ES2019 target and eventually move to ES2020 without the need to transpile down was also a factor. As it stands, the lack of adequate polyfill capabilities for CSS features in tandem with the goals of our partners is the biggest detriment to the support matrix not being higher.
+
+| Edge | Firefox | Chrome | Safari | Opera | Internet Explorer |
```
[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: Added these mobile browsers to the matrix

**Comment by Hotell** [comment]
> 

I'd suggest to setup a support window matrix/roadmap. Good example of what Im referring to can be seen here https://github.com/DefinitelyTyped/DefinitelyTyped#support-window

File: rfcs/shared/build-system/06-browser-support-for-v9.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# RFC: Browser Support for Fluent v9
+
+Contributors: Fluent UI React Build team, @TristanWatanabe
+
+## Summary
+
+Fluent v9 is currently lacking a defined browser support matrix so this RFC proposes a set of major browser versions compatible with the library to act as the source of truth. As mentioned below, Fluent developers are expected to conform to these browsers and ensure no incompatible syntax, browser APIs or CSS features are introduced to the library. Consumers who require older browser support that aren't covered in this matrix are responsible for performing downlevel transpilations and/or using polyfills themselves.
```
[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:7]: This was intended to say that Fluent wont be **providing** polyfills for features that we expect our browser matrix to already support (i.e. ResizeObserver) - i'll rephrase it

File: rfcs/shared/build-system/06-browser-support-for-v9.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+### Browser Support Going Forward
+
+With an eye towards being a modern library, Fluent will follow a yearly audit process to evaluate and update the current browser support matrix to keep pace with the ever evolving ecosystem of the web. This will allow the team to utilize modern tools and improve overall engineering efficiency while also providing consumers with performance and bundlesize improvements. Due to the nature of how Safari browser updates are tied to MacOS versions, Safari will be the browser that ultimately holds the library back from fully adopting modern browser versions across the board. Because of this, the proposed timeframe for when these audits occur should be whenever a Safari major release occurs. These tend to happen annually during the month of September so Fluent can target an annual audit around that time.
```
[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:39]: I think this auditing process will be heavily influenced by partner browser support needs so yearly audits seemed like a reasonable timeframe to evaluate, decide and give notice - totally open to suggestions here though!

[the remark this answers is not in the document]
Comment (Hotell) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:39]: I'd suggest to remove the "blaming" on safari and keep the wording as you mentioned - yearly audits etc. wdyt ?

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (layershifter) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: nit: I would have mobile browsers in a separate table.

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:7]: I see, what I was trying to point out is that we cannot guarantee this as it cannot be baked by tooling to 100%. So we might end up shipping "some" polyfills.

To give you better context, based on this RFC our build process will look like following:

1. browserlist config with browser matrix
2. babel consuming browserlist and transpiling according to that 

Now one can use ECMA features beyond 2019 and it will just work because babel will transpile that + provide according polyfill from corejs

What about TypeScript? 

well we will still use `target` and `lib` specification - which will be used only for type declarations. Now for production code this might become problematic as target/lib cannot be matched to browser matrix in any way, so we'll end up by "approximate" target boundaries specification on type safety level.

File: rfcs/shared/build-system/06-browser-support-for-v9.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+As it stands, Fluent v9 does not currently adhere to a specific browser support matrix. This makes it difficult for partner teams and general consumers of the library to know how using the library will impact their own product. There's also a lack of guidance for the internal team in terms of the syntax, browser APIs and CSS features that they can/can't use.
+
+## Detailed Design or Proposal
+
+### Browser Support Matrix
+
+The proposed browser support matrix will be loosely based off the [Resize Observer](https://caniuse.com/mdn-api_resizeobserver) support matrix . This decision was driven by the current usage of the `ResizeObserver` in Fluent v9 as well as the current business goals of our partner teams. Along with that, being able to maintain our current ES2019 target and eventually move to ES2020 without the need to transpile down was also a factor. As it stands, the lack of adequate polyfill capabilities for CSS features in tandem with the goals of our partners is the biggest detriment to the support matrix not being higher.
+
+| Edge | Firefox | Chrome | Safari | Opera | Safari on IOS | Chrome for Android | Samsung | Internet Explorer |
```
Comment (Hotell) [inline rfcs/shared/build-system/06-browser-support-for-v9.md:19]: wondering about these 2 chrome versions. the version difference is quite big.

image

AFAIR chrome on android is "on par" with desktop v8 core. If that's true this doesn't make much sense and mobile version should be lowered. WDYT?

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:39]: Sounds good - i’ll rephrase this too

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 5 above)
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:19]: Ah you’re right, they are on par - i’ll update the mobile version to 80 as well

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (spmonahan) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: Don't really feel strongly about it either way but I'm curious why you want to put mobile browsers in a separate table?

[the remark this answers is not in the document]
Comment (layershifter) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: @spmonahan from my POV it will have more clarity as "Samsung" looks strange near to IE. But I don't have a strong opinion about it.

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: change made [2130194](https://github.com/microsoft/fluentui/pull/22889/commits/2130194582480fdb8da9fb4ede81e41e58af18d4)

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: Some rephrasing regarding this [ebe76cd](https://github.com/microsoft/fluentui/pull/22889/commits/ebe76cd1b5450a997404a4750fe0147d83035e70)

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md:21]: Updated to separate desktop/mobile support matrices: [9b081a0](https://github.com/microsoft/fluentui/pull/22889/commits/9b081a06ededa790a9d2804f3b083d5b611a622f)

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 1 above)
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: issue link for this: #23020

**Comment by TristanWatanabe** [comment]
Updated the proposed browser matrix to `flex gap` (1f25ef7 and fc05585)  to allow this RFC to get merged in since we agreed to address that issue post v9 stable release. Once ongoing partner convos are resolved and if we do end up having to remove `flex gap` usage from our components, our browser matrix can be updated again. It ultimately wont affect consumers if we lower our matrix since that just means we're supporting more browsers; the work in that case falls squarely on us to remove incompatible features from the library.

File: rfcs/shared/build-system/06-browser-support-for-v9.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (brandonthomas) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: We're seeing some of these same things pop up in MADS (admin-controls) as well in terms of `target` and `lib`. I like the line drawn. We also use nullish coalescing so I think ES2020 is our line. Do you plan to look at that as well? It's very helpful to avoid pesky falsey bugs when looking for values.

[the remark this answers is not in the document]
Comment (TristanWatanabe) [reply rfcs/shared/build-system/06-browser-support-for-v9.md]: @brandonthomas We definitely want to support ES2020 features like `nulling coalescing` which is why I ensured this browser matrix supports all ES2020 features so no downlevel transpilation need to occur when we do move our target to ES2020. So in essence, this does draw the line at ES2020 for now

**Review by Hotell — APPROVED** [review]
awesome work @TristanWatanabe 🙌

**Comment by brandonthomas** [comment]
@TristanWatanabe one other thing I forgot to mention that I think is important to consider here is CSS's logical properties. It makes positioning for things like RTL, BiDi, etc much easier to deal with and we no longer need to do things like flip positions/margins etc. I'm starting to use these in the MADS library now that support for them is quite high.

https://caniuse.com/css-logical-props
